llvm/lib/Target/X86/X86TargetTransformInfo.cpp

```diff
@@ -5157,8 +5157,9 @@ InstructionCost X86TTIImpl::getMemoryOpCost(unsigned Opcode, Type *Src,
 
   InstructionCost Cost = 0;
 
-  // Add a cost for constant load to vector.
-  if (Opcode == Instruction::Store && OpInfo.isConstant())
+  // Add a cost for constant load to vector, if pointer is not a constant.
+  if (auto *SI = dyn_cast_or_null<StoreInst>(I);
+      SI && !isa<Constant>(SI->getPointerOperand()) && OpInfo.isConstant())
```

Comment:
We're trying to account for the cost of the constant vector load, but you want to skip it if the store address is constant?

Reply:
Looking at #111126 - shouldn't we be checking to see if `OpInfo.isUniform()` and adjusting the Type accordingly?

It looks like the test case thinks it has to load a `<20 x i64> splat(i64 1)` constant when it just has to load `<2 x i64> splat(i64 1)`

Reply:
Sorry if I'm missing something, where is the Type being `<20 x i64>` in this case? AFAICT, there isn't any load associated to the store when the store address is constant in x86, as it should just be needed to compute the memory address.

Reply:
The cost you are disabling in the code above is for the constant load - not the memory address computation.

Reply:
That's correct. There is no cost for any constant load. There is only a memory address computation, which is _not_ a load.

Reply:
That's only true for scalars - for vector constant it must be loaded from the constant pool unless rematerialized (which for x86 means 0 or -1 splats only). That is the issue in #111126 - the load from `LCPI0_0` is being overestimated (its just a <2 x i64> load that is reused in all 10 stores but we're treating it as a <20 x i64> load):
```
f():
        movaps  xmm0, xmmword ptr [rip + .LCPI0_0]
        movaps  xmmword ptr [rip + arr], xmm0
        movaps  xmmword ptr [rip + arr+16], xmm0
        movaps  xmmword ptr [rip + arr+32], xmm0
        movaps  xmmword ptr [rip + arr+48], xmm0
        movaps  xmmword ptr [rip + arr+64], xmm0
        movaps  xmmword ptr [rip + arr+80], xmm0
        movaps  xmmword ptr [rip + arr+96], xmm0
        movaps  xmmword ptr [rip + arr+112], xmm0
        movaps  xmmword ptr [rip + arr+128], xmm0
        movaps  xmmword ptr [rip + arr+144], xmm0
        ret
```
I'm investigating an alternative patch but have run out of time tonight :/

Reply:
Sorry, isn't this just bad codegen? In principle, we should be able to simply store the constant in the floating-point register, why would we need to load it from memory? GCC seems to catch this: https://godbolt.org/z/6baWMK4cW.

Reply:
Uhmm, we still need a load when dealing with arbitrary 64-bit constants.

Reply:
If any, the check would need to take this into account (scalar <= INT_MAX), but feel free to assess if there's a more favourable approach here.